setversion.py

```diff
@@ -0,0 +1,120 @@
+def changelog_rewriter(line, ctxt):
+    expr = ctxt.get("changelog_expr", None)
+    if not expr:
+        expr = re.compile(r"^\.\. top-of-changelog")
+        ctxt["changelog_expr"] = expr
+    if expr.match(line):
+        datestr = date.today().strftime("%B %d, %Y")
+        cl1 = "CbAPI {0} - Released {1}".format(ctxt["version"], datestr)
+        cl2 = "".ljust(len(cl1), "-")
+        updates = "\n\nUpdates\n\n.. (add your updates here)\n"
+        return line + "\n" + cl1 + "\n" + cl2 + updates
+    return None
+
+
+def doc_conf_rewriter(line, ctxt):
+    vexpr = ctxt.get("doc_version_expr", None)
+    if not vexpr:
+        vexpr = re.compile(r"^version = ")
+        ctxt["doc_version_expr"] = vexpr
+    rexpr = ctxt.get("doc_release_expr", None)
+    if not rexpr:
+        rexpr = re.compile(r"^release = ")
+        ctxt["doc_release_expr"] = rexpr
+    if vexpr.match(line):
+        t = re.match(r"^(\d+\.\d+)\.", ctxt["version"])
+        vn = ctxt["version"]
+        if t:
+            vn = t.group(1)
+        return "version = u'{0}'\n".format(vn)
+    if rexpr.match(line):
+        return "release = u'{0}'\n".format(ctxt["version"])
+    return None
+
+
+def setup_rewriter(line, ctxt):
+    expr = ctxt.get("setup_expr", None)
+    if not expr:
+        expr = re.compile(r"^(\s*)version=")
+        ctxt["setup_expr"] = expr
+    m = expr.match(line)
+    if m:
+        return "{0}version='{1}'\n".format(m.group(1), ctxt["version"])
+    return None
+
+
+def init_rewriter(line, ctxt):
+    expr = ctxt.get("init_expr", None)
+    if not expr:
+        expr = re.compile(r"^__version__ = ")
+        ctxt["init_expr"] = expr
+    if expr.match(line):
+        return "__version__ = '{0}'\n".format(ctxt["version"])
+    return None
+
+
+def rewrite_file(infilename, rewritefunc, ctxt):
+    outfilename = infilename + ".new"
+    infile = open(infilename, "r")
+    outfile = open(outfilename, "w")
+    try:
+        s = infile.readline()
+        while s:
+            s2 = rewritefunc(s, ctxt)
+            if s2:
+                outfile.write(s2)
+            else:
+                outfile.write(s)
+            s = infile.readline()
+    finally:
+        infile.close()
+        outfile.close()
+    if not ctxt["nodelete"]:
+        os.remove(infilename)
+        os.rename(outfilename, infilename)
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Set the version number in CbAPI source and documentation.\n"
+                                                 "Execute this on a release or hotfix branch to update "
+                                                 "the version numbers in the source.",
+                                     epilog="After running, edit docs/changelog.rst and add the new changelog "
+                                            "information under the new heading.")
+    parser.add_argument("version", help="New version number to add")
+    parser.add_argument("-n", "--nodelete", action="store_true",
```

Comment:
I see the benefit to creating the .new files though does it make sense to add another action to move .new to the original name that way the user doesn't have to go through and rename all 5 files?

Reply:
This would be a follow up call to the script after reviewing of course

Reply:
Not necessary.  If you rerun the script without the -n flag but with the same version number, the .new files will get rewritten, and then moved over the existing files.

I _could_ add an option to keep the old files around instead of deleting them, though, and I think I will.

Reply:
The way I see the process flow in my head is 
1. You run the script with the new version and generate the .new files
2. You review the output and add the necessary text to the changelog.rst file
3. Once satisfied with the results, run the script with a different command to accept the .new files and overwrite the original
4. Commit and push

Reply:
I added the option --renameonly or -r to support your desired workflow. This does make the script quite a bit more complicated though.

Reply:
I didn't mean an optional property. But instead something like this `{script} accept`

Reply:
_Now_ it should work like you want it to.